proposals/2180-search-service.md

```diff
@@ -0,0 +1,72 @@
+ * Within current Marquez architecture it is really important to have service up and running, so that incoming lineage
+events get saved in database. We don't want dummy search query to stop loading the events.
+ * We cannot create TEXT index on all the text fields. Creating index on jsonb columns could be dangerous as we know the 
+size of such fields can be counted in megabytes. 
+
+[ElasticSearch](https://www.elastic.co) is widely used open-source search engine. It is beneficial to have its capabilities in Marquez because:
+ * It is designed to index and search JSON documents and OpenLineage events are JSONs by definition. 
+ * It is capable of searching by selected JSON fields or a whole OpenLineage event.
+ 
+### Pluggable Marquez Search Engine
+
+Some users find it extremely useful to use Marquez as simple UI for browsing OpenLineage events that is really easy
+easy to install. For those ElasticSearch can be an overkill. We still want to have an option to run Marquez without Search
+Service which will mean limited search capabilities but with simple installation steps.
+
+Similarly to database configuration, `yaml` configuration could contain extra section:
+```
+searchService:
+  enabled: true
+  elastic.url: http://my-elastic-cluster:9200
+```
+
+Marquez CLI should be capable of indexing whole `lineage_events` table within ElasticSearch indexes.
+
+Whenever lineage event is received in Marquez, it is sent to ElasticSearch indexes. Keep in mind that assuring uniqueness on
+Elastic side may be difficult, so in case of sending multiple same events Marquez will have a single copy of that (as long
+as it's a single run id), but SearchService will contain duplicate entries. 
+
+New `search` endpoint shall be implemented in Marquez and available only when `searchService` enabled.
+
+### Search Service in action
+
+ * Two indexes in ElasticSearch (ES) need to be created:
+   * `lineage`: to contain whole lineage events,
+   * `dataset`: to contain single elements of `inputs` and `outputs`.
+ * Elastic search returns found documents, so separate `dataset` index is required. Otherwise, when searching for datasets, 
+returning whole OpenLineage event would not be sufficient. 
+ * A user sends request to `/api/v1/search` with a query specified. Additional filter may be used to specify if `dataset`, `job` or both are to be searched. 
+ * Current implementation returns instances of `SearchResult` class containing:
+ ```
+  ResultType type; // enum DATASET or JOB
+  String name;
+  Instant updatedAt;
+  NamespaceName namespace;
+  NodeId nodeId;
+```
+ * In order to find datasets ES will query `dataset` index. When searching for jobs, `run` and `job` subtrees of `lineage` index will be queried. 
+ * A response from ES will contain runId for jobs, and (namespace, name) tuples for datasets. 
+ * Based on that, additional query to PostgreSQL will be sent to convert this into list of `SearchResult` entries.
```

Comment:
What would be returned by querying postgres? Wouldn't the object be stored in ES and returned to the caller?

Reply:
right, it's not necessary. `SearchResult` can be evaluated from ElasticSearch backend.

Reply:
I think as a follow up, we can discuss how we might want to revisit the `SearchResult` model and was to extend it.